04_Scheduling/03_Tasks_And_Threads.md

```diff
@@ -0,0 +1,132 @@
+# Tasks and Threads
+
+** THIS SECTION IS IN EARLY STAGES For now it will be more a set of bullet lists with things to be expaneded in the future ** 
+
+## Tasks
+
+### What are tasks
+
+In a task usually there are the following information: 
+
+* Virtual memory base
+* Task ID
+* Thread List (if we implement multithreading) or function pointer if we don't
+* Stack reference
+* Arguments 
+* Parent task
+
+While a thread should contains at minimum:
+
+* Thread name
+* Thread id
+* Stack pointer
+* Execution frame
+* Pointer to function to run
+* Eventually arguments for the function
+* Depending on how the scheduler is designed, it could contains a pointer to the next thread
+* Task containing the current thread
+
+### Prerequisites 
+
+Even if technically not necessary, is better to have implemented a memory allocation mechanism, ideally both physical and virtual memory manager should be implemented, if we want to have full tasks/thread separation and protection. But in case we don't want to have memory separation and protection in place a physical memory manager can be enough.
+
+Technically we could also decide to implement a "kind of" multitasking without any memory allocator, but in this case we must use an array with a fixed size. So the Operating system will be limited to a certain number of concurrent task, and this memory will be always be unavailable to the os even when the tasks are dead.
+
+### Creating a thread
+
+When creating a thread we must first of all create it's execution frame (yeah even if not executed yet, it will need one).
+
+### Execution frame
+
+One of the way to implement stack switching is replacing the current iret frame with a the one contained in the next thread. To do that a task need when created to build an ad-hoc frame that will not cause any crash. What it should contain? Well this is basically a copy of the stack frame during the handling of an interrupt routine. 
+
+### What a task should contain ?
+
+So what are the information that we should store in a task? That depends on the design choices, because if we are implementing a multitask/single thread os it will need more information, but if we are going to implement a multi-task/multi-threaded OS some of the information will be stored in the Thread structure. So let's start with the minimum set of information that we need to save into a task: 
+
+* First of all a name, all tasks have a task name, most of the time is the excuatble name, sometime something different
```

Comment:
Task name is nice to have, but definitely not required info to store about a task.